specutils/spectra/xrayspectrum1d.py

```diff
@@ -0,0 +1,573 @@
+        tlmin = np.int(list(hdr.items())[tlmin_idx][1])
+
+        return tlmin
+
+    def _flatten_arrays(self, n_grp, f_chan, n_chan, matrix):
+
+        if not len(n_grp) == len(f_chan) == len(n_chan) == len(matrix):
+            raise ValueError("Arrays must be of same length!")
+
+        # find all non-zero groups
+        nz_idx = (n_grp > 0)
+
+        # stack all non-zero rows in the matrix
+        matrix_flat = np.hstack(matrix[nz_idx])
+
+        # stack all nonzero rows in n_chan and f_chan
+        #n_chan_flat = np.hstack(n_chan[nz_idx])
+        #f_chan_flat = np.hstack(f_chan[nz_idx])
+
+        # some matrices actually have more elements
+        # than groups in `n_grp`, so we'll only pick out
+        # those values that have a correspondence in
+        # n_grp
+        f_chan_new = []
+        n_chan_new = []
+        for i,t in enumerate(nz_idx):
+            if t:
+                n = n_grp[i]
+                f = f_chan[i]
+                nc = n_chan[i]
+                if np.size(f) == 1:
+                    f_chan_new.append(f)
+                    n_chan_new.append(nc)
+                else:
+                    f_chan_new.append(f[:n])
+                    n_chan_new.append(nc[:n])
+
+        n_chan_flat = np.hstack(n_chan_new)
+        f_chan_flat = np.hstack(f_chan_new)
+
+        # if n_chan is zero, we'll remove those as well.
+        nz_idx2 = (n_chan_flat > 0)
+        n_chan_flat = n_chan_flat[nz_idx2]
+        f_chan_flat = f_chan_flat[nz_idx2]
+
+        return n_grp, f_chan_flat, n_chan_flat, matrix_flat
+
+    def apply_rmf(self, spec):
+        """
+        Fold the spectrum through the redistribution matrix.
+
+        The redistribution matrix is saved as a flattened 1-dimensional
+        vector to save space. In reality, for each entry in the flux
+        vector, there exists one or more sets of channels that this
+        flux is redistributed into. The additional arrays `n_grp`,
+        `f_chan` and `n_chan` store this information:
+        
+            * `n_group` stores the number of channel groups for each
+              energy bin
+
+            * `f_chan` stores the *first channel* that each channel
+              for each channel set
+
+            * `n_chan` stores the number of channels in each channel
+              set
+
+        As a result, for a given energy bin i, we need to look up the
+        number of channel sets in `n_grp` for that energy bin. We
+        then need to loop over the number of channel sets. For each
+        channel set, we look up the first channel into which flux
+        will be distributed as well as the number of channels in the
+        group. We then need to also loop over the these channels and
+        actually use the corresponding elements in the redistribution
+        matrix to redistribute the photon flux into channels.
+
+        All of this is basically a big bookkeeping exercise in making
+        sure to get the indices right.
+
```

Comment:
In fact, it's so painful as a bookkeeping exercise that I won't do it by hand now. That's why it's so important to test with at least one real-world rmf in the tests below.